posts/2021/11/pydata-extensibility-vision.md

```diff
@@ -0,0 +1,388 @@
+
+# In CuPy codebase
+# from scipy.ndimage import laplace
+import cupyx.scipy.ndimage as _cupyx_ndimage
+
+@laplace.register(cupy.ndarray)
+def cupy_laplace(input, output=None, mode="reflect", cval=0.0):
+    return _cupyx_ndimage.laplace(input, output=output, mode=mode, cval=cval)
+
+# run registration code when CuPy is imported
+```
+
+We begin by defining the base `laplace` function, it is a fallback
+implementation that is going to be called for non-registered types of `input`.
+Next `laplace.register` is used to connect the implementation that is specific
+to `cupy.ndarray` with the base `laplace` function. We get a function in the
+`scipy.ndimage` module that is now compatible with CuPy arrays. Similarly, it
+can be extended to work with Dask arrays.
+
+```python
+from scipy.ndimage import laplace
+import numpy as np
+import cupy as cp
+
+numpy_img = np.random.normal(size=(256, 256))
+cupy_img = cp.array(numpy_img)
+
+print(type(laplace(numpy_img)))
+# <class 'numpy.ndarray'>
+
+print(type(laplace(cupy_img)))
+# <class 'cupy._core.core.ndarray'>
+```
+
+The above single-dispatch example illustrates the dispatching principle, but is
+too simple for our purposes - so let's look at multiple dispatch methods.
+The Python stdlib doesn't provide tools for multiple dispatch, however there
+are several dedicated libraries:
+
+* [multimethod](https://github.com/coady/multimethod) - pure Python
+  implementation of multiple dispatch with caching of argument types
+* [Plum](https://github.com/wesselb/plum) - implementation of multiple
+  dispatch that follows the ideas from Julia
+* [multipledispatch](https://github.com/mrocklin/multipledispatch) - similar
+  to Plum, only with slightly fewer features and no longer developed; vendored
+  in SymPy
+* [uarray](https://github.com/Quansight-Labs/uarray) is a generic
+  backend/multiple dispatch library similar to NumPy's dispatch functionality
+  except that the implementation doesn't need to be baked into the array type;
+  used in SciPy's `fft` module.
+
+The first three options are very similar in functionality and usability. But the last
+option, in addition to multiple dispatch, offers granular control using context
+managers and provides a way to switch backends for the same array type. It's
+also possible to make the `uarray` dispatcher work without a context manager
+making the mechanism implicit and similar to other dispatch libraries. Whether
+we should encourage the implicit registration of backends or not is an open
+design issue and there's previous discussion on the topic in
+[SciPy #14266](https://github.com/scipy/scipy/issues/14266). Hameer Abbasi, one of the
+`uarray` authors, wrote
+[a blog post](https://labs.quansight.org/blog/2019/07/uarray-update-api-changes-overhead-and-comparison-to-__array_function__/)
+about the motivation for `uarray` and how it compares to NumPy's
+`__array_function__` dispatch mechanism.
+
+## It's already happening
```

Comment:
I'm inclined to change this title to `Next steps`. Most of the content below is already reflected higher up. What we should end with I think is a rough outline of the next steps:
- discuss on Discourse to align with maintainers of SciPy, scikit-learn, scikit-image and the wider community;
- come to a conclusion on a few key points, in particular the opt-in vs. opt-out topic
- merge some of the uarray backends, and then explore using the array API standard in SciPy and scikit-learn with CuPy 10.0

And then a concluding statemen, like "This is an exciting project, we look forward to ... <etc."

Reply:
Right, listing the next steps is a better way to end this blog post.

Reply:
@IvanYashchuk will have a go at this one.

Reply:
Added in https://github.com/Quansight-Labs/quansight-labs-site/pull/278/commits/2e77cdc415c9df34d52d4be478b9e3335401f91d.
Once we have direct links to the discussions we should update the text here.

Reply:
Looks good too. Pushed some tiny tweaks, overall it read quite well.